protocols/gossipsub/src/types.rs

```diff
@@ -208,46 +252,68 @@ pub struct PeerInfo {
-        /// The topic of the messages.
-        topic_hash: TopicHash,
-        /// A list of known message ids (peer_id + sequence _number) as a string.
-        message_ids: Vec<MessageId>,
-    },
+    IHave(IHave),
     /// The node requests specific message ids (peer_id + sequence _number) - IWant control message.
-    IWant {
-        /// A list of known message ids (peer_id + sequence _number) as a string.
-        message_ids: Vec<MessageId>,
-    },
+    IWant(IWant),
     /// The node has been added to the mesh - Graft control message.
-    Graft {
-        /// The mesh topic the peer should be added to.
-        topic_hash: TopicHash,
-    },
+    Graft(Graft),
     /// The node has been removed from the mesh - Prune control message.
-    Prune {
-        /// The mesh topic the peer should be removed from.
-        topic_hash: TopicHash,
-        /// A list of peers to be proposed to the removed peer as peer exchange
-        peers: Vec<PeerInfo>,
-        /// The backoff time in seconds before we allow to reconnect
-        backoff: Option<u64>,
-    },
+    Prune(Prune),
 }
 
-/// A Gossipsub RPC message sent.
+/// Node broadcasts known messages per topic - IHave control message.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct IHave {
+    /// The topic of the messages.
+    pub(crate) topic_hash: TopicHash,
+    /// A list of known message ids (peer_id + sequence _number) as a string.
+    pub(crate) message_ids: Vec<MessageId>,
+}
+
+/// The node requests specific message ids (peer_id + sequence _number) - IWant control message.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct IWant {
+    /// A list of known message ids (peer_id + sequence _number) as a string.
+    pub(crate) message_ids: Vec<MessageId>,
+}
+
+/// The node has been added to the mesh - Graft control message.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Graft {
+    /// The mesh topic the peer should be added to.
+    pub(crate) topic_hash: TopicHash,
+}
+
+/// The node has been removed from the mesh - Prune control message.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Prune {
+    /// The mesh topic the peer should be removed from.
+    pub(crate) topic_hash: TopicHash,
+    /// A list of peers to be proposed to the removed peer as peer exchange
+    pub(crate) peers: Vec<PeerInfo>,
+    /// The backoff time in seconds before we allow to reconnect
+    pub(crate) backoff: Option<u64>,
+}
+
+/// A Gossipsub RPC message sent.
+#[derive(Debug)]
 pub enum RpcOut {
-    /// Publish a Gossipsub message on network.
-    Publish(RawMessage),
-    /// Forward a Gossipsub message to the network.
-    Forward(RawMessage),
+    /// Publish a Gossipsub message on network. [`Delay`] tags the time we attempted to send it.
+    Publish { message: RawMessage, timeout: Delay },
+    /// Forward a Gossipsub message on network. [`Delay`] tags the time we attempted to send it.
+    Forward { message: RawMessage, timeout: Delay },
```

Comment:
I find the current description a bit confusing, Wdyt think of the below, based on `Config::{forward, publish}_queue_duration`

```suggestion
    /// Publish a Gossipsub message on network.`timeout` limits the duration the message 
    /// can wait to be sent before it is abandoned.
    Publish { message: RawMessage, timeout: Delay },
    /// Forward a Gossipsub message on network. `timeout` limits the duration the message 
    /// can wait to be sent before it is abandoned.
    Forward { message: RawMessage, timeout: Delay },
```